File: sc_song.py
import base64
import re
import socket
import time

# Configuration
scip = "your_shoutcast_server_ip"
scport = 8000
scpass = "your_shoutcast_server_admin_password"

# Connect to shoutcast server
try:
    scfp = socket.create_connection((scip, scport), timeout=30)
except ConnectionRefusedError:
    print("Shoutcast server is offline")
    exit()

# Authenticate and retrieve XML data
auth_header = "GET /admin.cgi?mode=viewxml HTTP/1.1\r\nHost: {}:{}\r\nUser-Agent: SHOUTcast Song (author: dstjohn@mediacast1.com)(Mozilla Compatible)\r\nAuthorization: Basic {}\r\n\r\n".format(
    scip, scport, base64.b64encode(f"admin:{scpass}".encode()).decode()
)
scfp.sendall(auth_header.encode())

data = b""
while True:
    chunk = scfp.recv(4096)
    if not chunk:
        break
    data += chunk

# Decode XML data
page = data.decode()

# Define XML elements
elements = {
    "STREAMSTATUS": None,
    "BITRATE": None,
    "SERVERGENRE": None,
    "SERVERTITLE": None,
    "SONGTITLE": None,
}

for element in elements.keys():
    match = re.search(f"<{element}>(.*?)</{element}>", page)
    if match:
        elements[element] = match.group(1)

# Print server status and song history
if elements["STREAMSTATUS"]:
    print("Server Genre:", elements["SERVERGENRE"])
    print("Server Title:", elements["SERVERTITLE"])
    print("Song Title:", elements["SONGTITLE"])
else:
    print("Shoutcast server is offline")

song_history = re.findall("<SONG>(.*?)</SONG>.*?<PLAYEDAT>(.*?)</PLAYEDAT>", page, re.DOTALL)
for index, (song, played_at) in enumerate(song_history, start=1):
    formatted_date = time.strftime("%A, %d %B %Y %I:%M:%S %p", time.localtime(int(played_at)))
    print(f"{index}. Song: {song} - Played @ {formatted_date}")

scfp.close()

File: sc_song.php
<?php

/**

@Author: dstjohn (Mediacast1/Casterclub)
@Date started: 05-03-2002 (10:00A.M)
@Date Ended: 05-03-2002 (6:03 P.M)
@Date Updated: 05-03-2002 (6:06 P.M) Almost forgot, nothing like a 20 year pull request in the making.
@Requirements:
1.SHOUTcast streaming server
2.Oddcast dsp with winamp/xmms (recomended setup)
3.Webserver with php 4.x (Recommended environment: Unix (Freebsd, Red Hat etc.. with Apache 3.x)
@Support: None, post in the casterclub forums
@Core script Information:
SHOUTcast Song Status was written and developed on Windows Xp with apache and php4.1.2
Has not been tested on IIs webservers, if you do so and get it to work please let us know
At the forums (http://casterclub.com/forums)
Also has been tested on freebsd with apache, php4.1.2 and works fine.
*/

//connect to shoutcast server
include('./config.php');  //you may edit this path to fit your server environment otherwise leave it alone
$scfp = fsockopen("$scip", $scport, &$errno, &$errstr, 30);
 if(!$scfp) {
  $scsuccs=1;
echo''.$scdef.' is Offline';
 }
if($scsuccs!=1){

  //for newer shoutcast servers
fputs ($scfp, "GET /admin.cgi?mode=viewxml HTTP/1.1\r\nHost: $scip:$scport\r\n .
User-Agent: SHOUTcast Song (author: dstjohn@mediacast1.com)(Mozilla Compatible)\r\n .
Authorization: Basic ".base64_encode ("admin:$scpass")."\r\n\r\n");
 while(!feof($scfp)) {
  $page .= fgets($scfp, 1000);
 }

//define  xml elements
 $loop = array("STREAMSTATUS", "BITRATE");
 $y=0;
 while($loop[$y]!=''){
  $pageed = ereg_replace(".*<$loop[$y]>", "", $page);
  $scphp = strtolower($loop[$y]);
  $$scphp = ereg_replace("</$loop[$y]>.*", "", $pageed);
  if($loop[$y]==SERVERGENRE || $loop[$y]==SERVERTITLE || $loop[$y]==SONGTITLE)
   $$scphp = urldecode($$scphp);

// uncomment the next line to see all variables
// echo'$'.$scphp.' = '.$$scphp.'<br>';
  $y++;
 }
//end intro xml elements

//get song info and history
 $pageed = ereg_replace(".*<SONGHISTORY>", "", $page);
 $pageed = ereg_replace("</SONGHISTORY>.*", "", $pageed);
 $songatime = explode("<SONG>", $pageed);
 $r=1;
 while($songatime[$r]!=""){
  $t=$r-1;
  $playedat[$t] = ereg_replace(".*<PLAYEDAT>", "", $songatime[$r]);
  $playedat[$t] = ereg_replace("</PLAYEDAT>.*", "", $playedat[$t]);
  $song[$t] = ereg_replace(".*<TITLE>", "", $songatime[$r]);
  $song[$t] = ereg_replace("</TITLE>.*", "", $song[$t]);
  $song[$t] = urldecode($song[$t]);


//format the date
$frmt_date[$t] = date('l dS of F Y h:i:s A',$playedat[$t]);
	
//you may edit the html below, make sure to keep variables intact
echo'
<b>'.$t.'.</b>Song:  '.$song[$t].' - <b>Played @</b> '.$frmt_date[$t].'<BR>
';


		$r++;
	}

fclose($scfp);
}
?>
